File: android/core-model/src/main/kotlin/com/rfmapper/core/model/Enums.kt
package com.rfmapper.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Which sensor produced an observation. Determines which optional fields are meaningful. */
@Serializable
enum class SensorType {
    WIFI_SCAN,
    WIFI_ASSOCIATION,
    BLE,
    RTT,
    GPS,

    /** An administrator-declared infrastructure sighting. A human assertion, not a measurement. */
    ZONE_ANCHOR,

    /** Administrator entry, e.g. "device physically seen here". */
    MANUAL,

    /** Synthesised from an external dataset; requires `metadata.import_source`. */
    IMPORT,
    ;

    val isRadioMeasurement: Boolean
        get() = this == WIFI_SCAN || this == WIFI_ASSOCIATION || this == BLE || this == RTT

    companion object {
        fun fromWireOrNull(value: String): SensorType? = entries.firstOrNull { it.name == value }
    }
}

/** What kind of identifier [Observation.radioIdentifier] holds. */
@Serializable
enum class IdentifierType {
    WIFI_BSSID,
    WIFI_SSID,
    BLE_MAC_PUBLIC,

    /**
     * A locally-administered (privacy-rotating) BLE address. Recorded, but never automatically
     * attributed to a device: see `docs/17-identity-and-attribution-policy.md`.
     */
    BLE_MAC_RANDOM,
    BLE_SERVICE_UUID,
    BLE_IBEACON,
    GNSS_FIX,
    OBSERVER_SELF,
    OTHER,
    ;

    /** True when this identifier is not a durable identity and must not be treated as one. */
    val isEphemeral: Boolean
        get() = this == BLE_MAC_RANDOM

    companion object {
        fun fromWireOrNull(value: String): IdentifierType? = entries.firstOrNull { it.name == value }
    }
}

@Serializable
enum class ObserverDeviceType {
    ANDROID_PHONE,
    ANDROID_TABLET,
    IOS_PHONE,
    IOS_TABLET,
    FIXED_OBSERVER,
    OTHER,
    ;

    companion object {
        fun fromWireOrNull(value: String): ObserverDeviceType? = entries.firstOrNull { it.name == value }
    }
}

@Serializable
enum class Platform {
    @SerialName("android")
    ANDROID,

    @SerialName("ios")
    IOS,

    @SerialName("other")
    OTHER,
}

/** Enrollment status of a managed device. */
@Serializable
enum class DeviceStatus {
    AUTHORIZED,
    INFRASTRUCTURE,
    BLOCKED,
    DISABLED,
    UNKNOWN,
}

@Serializable
enum class InfrastructureType {
    WIFI_AP,

    /**
     * Coarse zone evidence only. A long-range router without verified ranging capability belongs
     * here: a published coverage radius is not a ranging accuracy.
     */
    ZONE_ANCHOR,

    /** Only valid once a genuine RTT range has actually been obtained from this node. */
    RTT_ANCHOR,
    BLE_ANCHOR,
    OBSERVER_PHONE,
    ROUTER,
    OTHER,
}

/**
 * How deep an inference the evidence supports. The engine emits the deepest tier the evidence
 * justifies and stops there.
 */
@Serializable
enum class PrecisionTier {
    /** Device is somewhere in the monitored environment. */
    SITE_PRESENCE,

    BUILDING,

    ZONE,

    /** Coordinates with honest uncertainty. */
    APPROXIMATE_POSITION,

    /** Requires genuine ranging evidence (RTT) in the supporting observation set. */
    PRECISION_RANGE,
    ;

    val hasCoordinates: Boolean
        get() = this == APPROXIMATE_POSITION || this == PRECISION_RANGE
}

@Serializable
enum class MovementState {
    STATIONARY,
    MOVING,
    ZONE_TRANSITION,
    LOST,
    REAPPEARED,
    UNCERTAIN,
}

@Serializable
enum class ZoneEventType {
    RF_ZONE_ENTER,
    RF_ZONE_EXIT,
    RF_ZONE_TRANSITION,
}

@Serializable
enum class TopologyStatus {
    ADJACENT,
    NON_ADJACENT,
    RESTRICTED,

    /** The zone pair has no authored edge: suspicious, but not declared impossible. */
    UNKNOWN_EDGE,
}

@Serializable
enum class ZoneEdgeType {
    DOOR,
    CORRIDOR,
    OUTDOOR_PATH,
    STAIRS,
    RESTRICTED,
    IMPOSSIBLE,
}

@Serializable
enum class ZoneKind {
    ROOM,
    AREA,
    CORRIDOR,
    THRESHOLD,
    OUTDOOR,
    PATH,
}

/** Promotion state of a fingerprint. Only an administrator may move CANDIDATE to GROUND_TRUTH. */
@Serializable
enum class FingerprintStatus {
    CANDIDATE,
    GROUND_TRUTH,
    RETIRED,
}

/** Distinguishes deliberate survey capture from ordinary collection. */
@Serializable
enum class SampleKind {
    ORDINARY,
    GROUND_TRUTH,
    ;

    companion object {
        fun fromWireOrDefault(value: String?): SampleKind =
            entries.firstOrNull { it.name == value } ?: ORDINARY
    }
}

/** Whether a Wi-Fi scan result came from the scan that just completed, or from the platform cache. */
@Serializable
enum class ResultFreshness {
    FRESH,
    CACHED,
    UNKNOWN,
}

/** Sensor capabilities an observer installation can actually produce. */
@Serializable
enum class ObserverCapability {
    WIFI_SCAN,
    WIFI_ASSOCIATION,
    BLE,
    RTT,
    GPS,
}

File: ios/Sources/RFMapperCore/Enums.swift
/// The wire enums, mirroring `core-model/Enums.kt`.
///
/// Every raw value is the Kotlin constant name verbatim. These strings appear in CSV cells, in JSON
/// and in the Python Lab's own enums, so a rename here is a breaking change to the file contract
/// rather than a local refactor.

public enum SensorType: String, Codable, CaseIterable, Sendable {
    case wifiScan = "WIFI_SCAN"
    case wifiAssociation = "WIFI_ASSOCIATION"
    case ble = "BLE"
    case rtt = "RTT"
    case gps = "GPS"

    /// An administrator-declared infrastructure sighting. A human assertion, not a measurement.
    case zoneAnchor = "ZONE_ANCHOR"

    /// Administrator entry, e.g. "device physically seen here".
    case manual = "MANUAL"

    /// Synthesised from an external dataset; requires `metadata.import_source`.
    case imported = "IMPORT"

    public var isRadioMeasurement: Bool {
        switch self {
        case .wifiScan, .wifiAssociation, .ble, .rtt: return true
        default: return false
        }
    }
}

/// What kind of identifier ``Observation/radioIdentifier`` holds.
public enum IdentifierType: String, Codable, CaseIterable, Sendable {
    case wifiBssid = "WIFI_BSSID"
    case wifiSsid = "WIFI_SSID"
    case bleMacPublic = "BLE_MAC_PUBLIC"

    /// A locally-administered (privacy-rotating) BLE address. Recorded, but never automatically
    /// attributed to a device: see `docs/17-identity-and-attribution-policy.md`.
    case bleMacRandom = "BLE_MAC_RANDOM"
    case bleServiceUuid = "BLE_SERVICE_UUID"
    case bleIBeacon = "BLE_IBEACON"
    case gnssFix = "GNSS_FIX"
    case observerSelf = "OBSERVER_SELF"
    case other = "OTHER"

    /// True when this identifier is not a durable identity and must not be treated as one.
    public var isEphemeral: Bool { self == .bleMacRandom }
}

public enum ObserverDeviceType: String, Codable, CaseIterable, Sendable {
    case androidPhone = "ANDROID_PHONE"
    case androidTablet = "ANDROID_TABLET"
    case iosPhone = "IOS_PHONE"
    case iosTablet = "IOS_TABLET"
    case fixedObserver = "FIXED_OBSERVER"
    case other = "OTHER"
}

public enum Platform: String, Codable, CaseIterable, Sendable {
    case android
    case ios
    case other
}

public enum DeviceStatus: String, Codable, CaseIterable, Sendable {
    case authorized = "AUTHORIZED"
    case infrastructure = "INFRASTRUCTURE"
    case blocked = "BLOCKED"
    case disabled = "DISABLED"
    case unknown = "UNKNOWN"
}

/// Distinguishes deliberate survey capture from ordinary collection.
public enum SampleKind: String, Codable, CaseIterable, Sendable {
    case ordinary = "ORDINARY"
    case groundTruth = "GROUND_TRUTH"

    public static func fromWireOrDefault(_ value: String?) -> SampleKind {
        guard let value, let parsed = SampleKind(rawValue: value) else { return .ordinary }
        return parsed
    }
}

/// Whether a scan result came from the scan that just completed, or from the platform cache.
public enum ResultFreshness: String, Codable, CaseIterable, Sendable {
    case fresh = "FRESH"
    case cached = "CACHED"
    case unknown = "UNKNOWN"
}

/// Sensor capabilities an observer installation can actually produce.
///
/// On iOS this set is smaller than on Android and the difference is load-bearing rather than
/// cosmetic: see `docs/06-ios-capability-matrix.md` §6.2.
public enum ObserverCapability: String, Codable, CaseIterable, Sendable {
    case wifiScan = "WIFI_SCAN"
    case wifiAssociation = "WIFI_ASSOCIATION"
    case ble = "BLE"
    case rtt = "RTT"
    case gps = "GPS"
}

public enum ExportKind: String, Codable, CaseIterable, Sendable {
    case day = "DAY"
    case session = "SESSION"
    case range = "RANGE"
}

public enum PackageType: String, Codable, CaseIterable, Sendable {
    case observations = "OBSERVATIONS"
    case derived = "DERIVED"
}
